### PR TITLE
v5 UDP packet handling

### PR DESCRIPTION
**Pull request by YouJustDontKnow**
UDP packet handling for discv5 protocol implementation

### CODE REVIEW

**Comment by atoulme** [comment]
ok to test this

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/PacketCodec.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.apache.tuweni.devp2p.v5
+
+import org.apache.tuweni.devp2p.v5.packet.UdpMessage
+import java.nio.ByteBuffer
+
+interface PacketCodec {
+
+  fun encode(message: UdpMessage): ByteBuffer
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/PacketCodec.kt]: Use Bytes, not ByteBuffer in interfaces

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package org.apache.tuweni.devp2p.v5
+
+import kotlinx.coroutines.Job
+import org.apache.tuweni.bytes.Bytes
+import org.apache.tuweni.devp2p.v5.packet.UdpMessage
+import java.net.InetSocketAddress
+
+interface UdpConnector {
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt]: Javadoc please

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/MessageDecoder.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package org.apache.tuweni.devp2p.v5
+
+import org.apache.tuweni.bytes.Bytes
+import org.apache.tuweni.crypto.SECP256K1
+import org.apache.tuweni.devp2p.v5.packet.UdpMessage
+
+interface MessageDecoder<T: UdpMessage> {
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/MessageDecoder.kt]: Javadoc - what is the point of exposing this class as an interface?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/NodeDiscoveryService.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package org.apache.tuweni.devp2p.v5
+
+interface NodeDiscoveryService {
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/NodeDiscoveryService.kt]: Javadoc please - what is the point of exposing this directly?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  fun start(): Job
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt]: Can we avoid coroutines in dependencies for the interface please

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  fun terminate()
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/UdpConnector.kt]: can we make this async?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
+package org.apache.tuweni.devp2p.v5.internal
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: Missing copyright headers

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.tuweni.devp2p.parseEnodeUri
+import org.apache.tuweni.devp2p.v5.NodeDiscoveryService
+import org.apache.tuweni.devp2p.v5.UdpConnector
+import org.apache.tuweni.devp2p.v5.packet.RandomMessage
+import org.apache.tuweni.io.Base64URLSafe
+import java.net.InetSocketAddress
+import java.time.Instant
+import kotlin.coroutines.CoroutineContext
+
+class DefaultNodeDiscoveryService(
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: can we move this out of an internal package and inline to the interface file, marked as internal class?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  private val enrSeq: Long = Instant.now().toEpochMilli(),
+  private val selfENR: Bytes = EthereumNodeRecord.toRLP(keyPair, enrSeq, emptyMap(), bindAddress.address, null, bindAddress.port),
+  private val nodeId: Bytes = Hash.sha2_256(selfENR),
+  private val connector: UdpConnector = DefaultUdpConnector(nodeId, bindAddress),
+  override val coroutineContext: CoroutineContext = Dispatchers.Default
+) : NodeDiscoveryService, CoroutineScope {
+
+  override fun start() {
+    connector.start()
+    launch { bootstrap() }
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: Unsafe - please make this async

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  override fun terminate() {
+    connector.terminate()
+  }
+
+  private fun bootstrap() {
+    bootstrapENRList.map {
+      val encodedEnr = it.substringAfter("enr:")
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: unguarded - what if the string is invalid?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.tuweni.devp2p.v5.NodeDiscoveryService
+import org.apache.tuweni.devp2p.v5.UdpConnector
+import org.apache.tuweni.devp2p.v5.packet.RandomMessage
+import org.apache.tuweni.io.Base64URLSafe
+import java.net.InetSocketAddress
+import java.time.Instant
+import kotlin.coroutines.CoroutineContext
+
+class DefaultNodeDiscoveryService(
+  private val keyPair: SECP256K1.KeyPair,
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: Why do we hardcode a SECP256K1 key pair for our own ENR? ENRs take any - consider taking an ENR as parameter instead.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      val encodedEnr = it.substringAfter("enr:")
+      val rlpENR = Base64URLSafe.decode(encodedEnr)
+      val enr = EthereumNodeRecord.fromRLP(rlpENR)
+      val destNodeId = Hash.sha2_256(rlpENR)
+
+      val randomMessage = RandomMessage(nodeId, destNodeId)
+      val address = InetSocketAddress(enr.ip(), enr.udp())
+
+      connector.addPendingNodeId(address, destNodeId)
+      connector.send(address, randomMessage)
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: Is this blocking?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultPacketCodec.kt [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    val content = bytes.slice(UdpMessage.TAG_LENGTH)
+    if (RLP.isList(content)) {
+      if (UdpMessage.isMagic(tag, nodeId)) {
+        return WhoAreYouMessage.create(content, nodeId, nodeId)
+      }
+    } else {
+      val srcId = UdpMessage.getSourceFromTag(tag, nodeId)
+      return RandomMessage.create(content, srcId, nodeId)
+    }
+    throw IllegalArgumentException("Unknown message inbound...")
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultPacketCodec.kt]: How is that exception caught and handled?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultUdpConnector.kt [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  override fun getPendingNodeIdByAddress(address: InetSocketAddress): Bytes = authenticatingPeers[address]
+    ?: throw IllegalArgumentException("Authenticated peer not found with address ${address.hostName}:${address.port}")
+
+  private fun processDatagram(datagram: ByteBuffer, address: InetSocketAddress) {
+    val message = packetCodec.decode(datagram)
+    when(message) {
+      is RandomMessage -> randomMessageHandler.handle(message, address, this)
+      is WhoAreYouMessage -> whoAreYouMessageHandler.handle(message, address, this)
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultUdpConnector.kt]: default?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultUdpConnector.kt [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+  private lateinit var receiveJob: Job
+
+  override fun start(): Job {
+    receiveChannel.bind(bindAddress)
+
+    receiveJob = launch {
+      val datagram = ByteBuffer.allocate(MAX_PACKET_SIZE)
+      while (receiveChannel.isOpen) {
+        datagram.clear()
+        val address = receiveChannel.receive(datagram) as InetSocketAddress
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultUdpConnector.kt]: What happens with the overflow bytes if any?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/handler/WhoAreYouMessageHandler.kt [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import org.apache.tuweni.bytes.Bytes
+import org.apache.tuweni.devp2p.v5.MessageHandler
+import org.apache.tuweni.devp2p.v5.UdpConnector
+import org.apache.tuweni.devp2p.v5.packet.WhoAreYouMessage
+import java.net.InetSocketAddress
+
+class WhoAreYouMessageHandler: MessageHandler<WhoAreYouMessage> {
+
+  override fun handle(message: WhoAreYouMessage, address: InetSocketAddress, connector: UdpConnector) {
+    println("Recieved WHOAREYOU message with auth tag ${message.authTag} and nonce ${message.idNonce}")
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/handler/WhoAreYouMessageHandler.kt]: consider replacing with the TODO Kotlin function.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/packet/UdpMessage.kt [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+  val dest: Bytes
+) {
+
+  protected val VERSION: Int = 5
+
+  abstract fun encode(): ByteBuffer
+
+  companion object {
+
+    const val MAX_UDP_MESSAGE_SIZE = 1280
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/packet/UdpMessage.kt]: this constant is defined twice?

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/packet/WhoAreYouMessage.kt [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+): UdpMessage(src, dest) {
+
+  override fun encode(): ByteBuffer {
+    val magic = magic(src)
+    val rlpBody = RLP.encodeList { w ->
+      w.writeByteArray(authTag.toArray())
+      w.writeByteArray(idNonce.toArray())
+    }
+    val messageBytes = Bytes.wrap(magic, rlpBody) // TODO: ENR seq number
+    return ByteBuffer.wrap(messageBytes.toArray())
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/packet/WhoAreYouMessage.kt]: yep, really should be returning Bytes here.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/EthereumNodeRecord.kt [hunk 19]
```diff
@@ -85,10 +86,10 @@ class EthereumNodeRecord(val signature: Bytes, val seq: Long, val data: Map<Stri
       }
       udp?.let {
         mutableData["udp"] = Bytes.ofUnsignedShort(it)
-        writer.writeString("udp")
       }
       mutableData.keys.sorted().forEach { key ->
           mutableData[key]?.let { value ->
+            println("$key -> $value")
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/EthereumNodeRecord.kt]: Please remove

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  fun encrypt(key: ByteArray, nonce: ByteArray, message: ByteArray, data: ByteArray): ByteArray {
+    val keySpec = SecretKeySpec(key, "AES")
+    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
+    val parameterSpec = GCMParameterSpec(128, nonce)
+
+    cipher.init(Cipher.ENCRYPT_MODE, keySpec, parameterSpec)
+
+    cipher.updateAAD(data)
+
+    val encriptedText = Bytes.wrap(cipher.doFinal(message))
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt]: encrypted, not encripted

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import javax.crypto.spec.SecretKeySpec
+
+object AES128GCM {
+
+  fun encrypt(key: Bytes, nonce: Bytes, message: Bytes, data: Bytes): Bytes {
+    val result = encrypt(nonce.toArray(), key.toArray(), data.toArray(), message.toArray())
+    return Bytes.wrap(result)
+  }
+
+  fun encrypt(key: ByteArray, nonce: ByteArray, message: ByteArray, data: ByteArray): ByteArray {
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt]: Please make it return a Bytes object.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    val nonceSize = Bytes.ofUnsignedInt(nonce.size.toLong())
+    return Bytes.wrap(nonceSize, wrappedNonce, encriptedText).toArray()
+  }
+
+  fun decrypt(encryptedContent: Bytes, key: Bytes, data: Bytes): Bytes {
+    val result = decrypt(encryptedContent.toArray(), key.toArray(), data.toArray())
+    return Bytes.wrap(result)
+  }
+
+  fun decrypt(encryptedContent: ByteArray, key: ByteArray, data: ByteArray): ByteArray {
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt]: Please make it return a Bytes object.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    val buffer = ByteBuffer.wrap(encryptedContent)
+    val nonceLength = buffer.int
+    val nonce = ByteArray(nonceLength)
+    buffer.get(nonce)
+    val encriptedText = ByteArray(buffer.remaining())
+    buffer.get(encriptedText)
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt]: encrypted

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt [hunk 24]
```diff
@@ -0,0 +1,56 @@
+package org.apache.tuweni.devp2p.v5.encrypt
```
Comment (atoulme) [inline devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/encrypt/AES128GCM.kt]: copyright header. Also need docs on class and each method.

File: devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt (hunk 11 above)
Comment (YouJustDontKnow) [reply devp2p/src/main/kotlin/org/apache/tuweni/devp2p/v5/internal/DefaultNodeDiscoveryService.kt]: Key pair is not for ENR dedicated. There is a private key required for secret agreement

**Pull request merged**